### PR TITLE
set sticky bit on connection files

### LINKED ISSUE #199: write_connection_file fails in long running notebook server
I'm on 64bit Linux, running `jupyter notebook` in a miniconda 3 environment (Python 3.5):

```
jupyter_client            4.3.0                    py35_0    conda-forge
jupyter_console           5.0.0                    py35_0    conda-forge
jupyter_core              4.1.0                    py35_0    conda-forge
```

When I leave the notebook server idle for extended periods (days, at least hours) I experience the following failure when creating a new notebook (Python 3 kernel):

```
[I 18:09:47.052 NotebookApp] Creating new notebook in
[E 18:09:48.373 NotebookApp] Unhandled error in API request
    Traceback (most recent call last):
      File ".../notebook/base/handlers.py", line 457, in wrapper
        result = yield gen.maybe_future(method(self, *args, **kwargs))
      File ".../tornado/gen.py", line 1015, in run
        value = future.result()
      File ".../tornado/concurrent.py", line 237, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
      File ".../tornado/gen.py", line 1021, in run
        yielded = self.gen.throw(*exc_info)
      File ".../notebook/services/sessions/handlers.py", line 62, in post
        kernel_id=kernel_id))
      File ".../tornado/gen.py", line 1015, in run
        value = future.result()
      File ".../tornado/concurrent.py", line 237, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
      File ".../tornado/gen.py", line 1021, in run
        yielded = self.gen.throw(*exc_info)
      File ".../notebook/services/sessions/sessionmanager.py", line 79, in create_session
        kernel_name)
      File ".../tornado/gen.py", line 1015, in run
        value = future.result()
      File ".../tornado/concurrent.py", line 237, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
      File ".../tornado/gen.py", line 1021, in run
        yielded = self.gen.throw(*exc_info)
      File ".../notebook/services/sessions/sessionmanager.py", line 92, in start_kernel_for_session
        self.kernel_manager.start_kernel(path=kernel_path, kernel_name=kernel_name)
      File ".../tornado/gen.py", line 1015, in run
        value = future.result()
      File ".../tornado/concurrent.py", line 237, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 3, in raise_exc_info
      File ".../tornado/gen.py", line 285, in wrapper
        yielded = next(result)
      File ".../notebook/services/kernels/kernelmanager.py", line 87, in start_kernel
        super(MappingKernelManager, self).start_kernel(**kwargs)
      File ".../jupyter_client/multikernelmanager.py", line 110, in start_kernel
        km.start_kernel(**kwargs)
      File ".../jupyter_client/manager.py", line 224, in start_kernel
        self.write_connection_file()
      File ".../jupyter_client/connect.py", line 406, in write_connection_file
        kernel_name=self.kernel_name
      File ".../jupyter_client/connect.py", line 134, in write_connection_file
        with open(fname, 'w') as f:
    FileNotFoundError: [Errno 2] No such file or directory: '/run/user/4150/jupyter/kernel-f440f50e-d0b9-4ea7-95bb-55300ef90602.json'
[E 18:09:48.650 NotebookApp] {
      "Content-Type": "application/json",
      "Content-Length": "76",
      "Dnt": "1",
      "Origin": "http://localhost:8888",
      "Connection": "keep-alive",
      "User-Agent": "...",
      "Accept-Encoding": "gzip, deflate",
      "Referer": "http://localhost:8888/notebooks/Untitled.ipynb?kernel_name=python3",
      "X-Requested-With": "XMLHttpRequest",
      "Host": "localhost:8888",
      "Accept": "application/json, text/javascript, */*; q=0.01",
      "Accept-Language": "en-GB,en;q=0.8,en-US;q=0.6,de;q=0.4,fr;q=0.2"
    }
[E 18:09:48.650 NotebookApp] 500 POST /api/sessions (::1) 481.34ms referer=http://localhost:8888/notebooks/Untitled.ipynb?kernel_name=python3
```

The issue turns out to be that the directory `/run/user/4150/jupyter` is missing. I suppose `/run/user/$UID` is cleaned up regularly by the system. Creating the directory or restarting the notebook server resolves the issue.

### PR DESCRIPTION
**Pull request by minrk**
avoids periodic cleanup of runtime directory

closes #199

### CODE REVIEW

**Comment by minrk** [comment]
Alternative: we could set the sticky bit on the runtime_dir itself.

**Comment by takluyver** [comment]
The spec only mentions setting the sticky bit on files, and whatever does the cleanup is probably running as root, so it wouldn't be bound by the sticky bit. So let's stick with doing it on the file rather than the directory.

File: jupyter_client/connect.py [hunk 1]
```diff
@@ -135,6 +136,21 @@ def write_connection_file(fname=None, shell_port=0, iopub_port=0, stdin_port=0,
     with open(fname, 'w') as f:
         f.write(json.dumps(cfg, indent=2))
 
+    if hasattr(stat, 'S_ISVTX'):
+        # set the sticky bit to avoid periodic cleanup
+        permissions = os.stat(fname).st_mod
```
Comment (takluyver) [inline jupyter_client/connect.py]: st_mode

**Comment by minrk** [comment]
The error in #199 suggests that the runtime_dir was cleaned up because writing a new file should only fail if the directory doesn't exist.

File: jupyter_client/connect.py (hunk 1 above)
Comment (minrk) [inline jupyter_client/connect.py]: Fixed, thanks

**Comment by minrk** [comment]
Now that I think about it, since #199 suggests that the directory is what was cleaned up, we may need something directory-related. Either ensure the directory exists when we try to write a connection file, or set the sticky bit on our runtime_dir.

**Comment by minrk** [comment]
Or do something non-sticky like call `os.utime` every once in a while to indicate that a given runtime file is still 'active'.

**Comment by takluyver** [comment]
Ah, right, we make a subdirectory inside the runtime dir. I'd forgotten that. Yes, in that case we probably need this sticky bit on the jupyter directory.

**Comment by minrk** [comment]
Finally got back to this one. It now sets the sticky bit on the directory containing the connection file, as well.

**Comment by AlJohri** [comment]
@minrk it seems like setting the sticky bit consistently fails on macOS because of a permission issue with the `/var` directory:

```
/Users/johria/.virtualenvs/heliograf-analysis-hss/lib/python3.6/site-packages/jupyter_client/connect.py:157: RuntimeWarning: Failed to set sticky bit on '/var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T': [Errno 1] Operation not permitted: '/var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T'
  RuntimeWarning,
```

**Comment by BoPeng** [comment]
I am seeing the same error on Mac.

```
/Users/bpeng1/anaconda/lib/python3.6/site-packages/jupyter_client/connect.py:157: 
RuntimeWarning: Failed to set sticky bit on '/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T': [Errno 1] 
Operation not permitted: '/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T'
```

**Comment by takluyver** [comment]
We're already catching the error and issuing a warning, which is what you see. Maybe we should skip the warning on Mac since it probably doesn't matter.

**Comment by BoPeng** [comment]
I frankly do not know if it matters but everything appears to be ok after the warning. The warning is pretty low level and scary so it is perhaps better to remove it (at least on mac).

**Comment by takluyver** [comment]
It shouldn't matter - setting the sticky bit is part of an XDG specification (XDG_RUNTIME_DIR), which as far as I know is only used on free desktop systems (Linux, BSD, etc.).

**Comment by AlJohri** [comment]
I run a command to rexecute a jupyter notebook in my Makefile and the output warning looks a bit scary for those unfamiliar. I'd give a +1 for either not setting the sticky bit on mac or suppressing the warning on macos.

**Comment by minrk** [comment]
@AlJohri Can you show the `ls -la` permissions on the directory? Are you its owner?

I'm okay skipping this on mac, but it's a bit surprising because I *don't* get an permission errors doing this, so I'd like to understand it, as well.

**Comment by AlJohri** [comment]
sure, I think `/var` is traditionally owned by root on macOS? I think I've run into this while trying to put some `pid` files in var run before. This particular path may be different though.

```
Als-MacBook-Pro # pwd
/var
Als-MacBook-Pro # ls -la
total 0
drwxr-xr-x  25 root          wheel          850 Jul 24 02:01 ./
drwxr-xr-x@  6 root          wheel          204 May 28 09:13 ../
drwx------   2 root          wheel           68 Jul 30  2016 agentx/
drwxr-xr-x   8 daemon        wheel          272 Sep 13  2016 at/
drwx------  25 root          wheel          850 Aug  8 11:15 audit/
drwx------   2 root          wheel           68 Jul 30  2016 backups/
drwxr-xr-x  70 root          wheel         2380 Aug  8 17:45 db/
drwxr-xr-x   2 root          sys             68 Jul 30  2016 empty/
drwxr-xr-x   4 root          wheel          136 Jul 20 11:28 folders/
drwxr-x---   2 _jabber       _jabber         68 Jul 30  2016 jabberd/
drwxr-xr-x   4 root          wheel          136 May 28 11:22 lib/
drwxr-xr-x  61 root          wheel         2074 Aug  9 08:29 log/
drwxr-x---   2 _mobileasset  _mobileasset    68 Jul 30  2016 ma/
drwxrwxr-x   2 root          mail            68 Jul 30  2016 mail/
drwxr-xr-x   3 root          wheel          102 Jul 30  2016 msgs/
drwxr-xr-x   2 root          wheel           68 Jul 30  2016 netboot/
drwxr-xr-x   5 _networkd     _networkd      170 May 28 09:16 networkd/
drwxr-x---   6 root          wheel          204 May 28 10:39 root/
drwxr-xr-x   4 root          wheel          136 Jul 30  2016 rpc/
drwxrwxr-x  31 root          daemon        1054 Aug  9 08:27 run/
drwxr-xr-x   2 daemon        wheel           68 Jul 30  2016 rwho/
drwxr-xr-x   6 root          wheel          204 Sep 13  2016 spool/
drwxrwxrwt   6 root          wheel          204 Aug  9 08:33 tmp/
drwxr-xr-x   4 root          wheel          136 Aug  8 12:22 vm/
drwxr-xr-x   3 root          wheel          102 Jul 30  2016 yp/
```

```
Als-MacBook-Pro # cd /var/folders/
Als-MacBook-Pro # ls -la
total 0
drwxr-xr-x   4 root  wheel  136 Jul 20 11:28 ./
drwxr-xr-x  25 root  wheel  850 Jul 24 02:01 ../
drwxr-xr-x@  3 root  wheel  102 May 28 09:27 c5/
drwxr-xr-x@ 17 root  wheel  578 May 28 20:17 zz/
```

```
Als-MacBook-Pro # pwd
/var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T
Als-MacBook-Pro # ls -lha
total 205304
drwx------@ 78 johria  staff   2.6K Aug  9 08:33 ./
drwxr-xr-x@  6 johria  staff   204B May 28 10:24 ../
drwx------   7 johria  staff   238B Aug  8 12:02 .AddressBookLocks/
drwx------   4 johria  staff   136B Aug  8 11:16 .BQR82RBBHL.com.tinyspeck.slackmacgap.NISTG4/
drwx------   2 johria  staff    68B Aug  8 11:15 .CalendarLocks/
drwx------@  4 johria  staff   136B Aug  8 11:15 .com.google.Chrome.dumnj0/
drwx------@  4 johria  staff   136B Aug  8 11:15 .io.nwjs.nw.KBzpRZ/
drwx------   2 johria  staff    68B Aug  8 11:15 .io.nwjs.nw.nFadVj/
-rw-r--r--   1 johria  staff     0B Aug  8 11:15 .keystoneAgentLock
drwx------@  2 johria  staff    68B Aug  8 11:15 2BUA8C4S2C.com.agilebits.onepassword-osx-helper/
drwxr-xr-x   2 johria  staff    68B Aug  8 13:24 MSau_10168/
-rw-------   1 johria  staff   559B Aug  8 11:59 SOSBackup-PCS-MasterKey-tomb
drwx------   3 johria  staff   102B Aug  8 23:53 SetupClient/
drwx------   2 johria  staff    68B Aug  9 08:33 TemporaryItems/
drwx------   3 johria  staff   102B Aug  8 11:15 assistantd/
drwx------@  3 johria  staff   102B Aug  8 11:15 chrome-1L4kxf/
drwx------@  3 johria  staff   102B Aug  8 11:15 chrome-J3KBSh/
drwx------@  2 johria  staff    68B Aug  8 11:47 com.agilebits.onepassword-osx.safariextensioncompanion/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.agilebits.onepasswordnativemessaginghost/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.AddressBook.ContactsAccountsService/
drwx------@  2 johria  staff    68B Aug  8 20:07 com.apple.AddressBook.FaceTimeService/
drwxr-xr-x   2 johria  staff    68B Aug  8 20:07 com.apple.AddressBook.thumbnailcache/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.AirPlayUIAgent/
drwx------@  3 johria  staff   102B Aug  8 11:15 com.apple.CalendarAgent/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.CalendarNotification.CalNCService/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.CloudPhotosConfiguration/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.ContactsAgent/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.MailCacheDelete/
drwx------@  4 johria  staff   136B Aug  8 18:16 com.apple.Notes/
drwx------@  2 johria  staff    68B Aug  8 11:17 com.apple.OSDUIHelper/
drwx------@  2 johria  staff    68B Aug  8 11:16 com.apple.PhotoIngestService/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.PressAndHold/
drwxr-xr-x@  3 johria  staff   102B Aug  8 11:47 com.apple.Safari/
drwx------@  3 johria  staff   102B Aug  8 11:15 com.apple.Safari.CacheDeleteExtension/
drwx------   2 johria  staff    68B Aug  8 11:16 com.apple.Safari.History/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.Siri/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.SocialPushAgent/
drwx------@  3 johria  staff   102B Aug  8 13:45 com.apple.WeatherKitService/
drwx------   2 johria  staff    68B Aug  8 11:15 com.apple.bird/
drwx------@  3 johria  staff   102B Aug  8 11:16 com.apple.cloudphotosd/
drwx------@  2 johria  staff    68B Aug  8 11:47 com.apple.corerecents.recentsd/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.ctkahp/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.dt.IDECacheDeleteAppExtension/
drwxr-xr-x   3 johria  staff   102B Aug  8 17:25 com.apple.finder/
drwx------@  3 johria  staff   102B Aug  8 11:47 com.apple.geod/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.iBooksX.CacheDelete/
drwx------@  2 johria  staff    68B Aug  8 16:06 com.apple.iCal/
drwx------@  2 johria  staff    68B Aug  8 13:45 com.apple.iCal.CalendarNC/
drwx------@  3 johria  staff   102B Aug  8 11:17 com.apple.iChat/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.iTunesCacheExtension/
drwx------@  2 johria  staff    68B Aug  8 17:46 com.apple.lateragent/
drwx------@  2 johria  staff    68B Aug  8 17:46 com.apple.mediaanalysisd/
drwx------@  2 johria  staff    68B Aug  8 13:45 com.apple.ncplugin.stocks/
drwx------@  2 johria  staff    68B Aug  8 13:45 com.apple.ncplugin.weather/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.notificationcenterui.WeatherSummary/
drwx------@  3 johria  staff   102B Aug  8 17:45 com.apple.photoanalysisd/
drwx------@  3 johria  staff   102B Aug  8 11:16 com.apple.photolibraryd/
drwx------@  3 johria  staff   102B Aug  8 11:16 com.apple.photomoments/
drwx------@  2 johria  staff    68B Aug  8 17:24 com.apple.quicklook.ui.helper/
drwx------@  2 johria  staff    68B Aug  8 17:44 com.apple.siri.media-indexer/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.apple.soagent/
drwx------   2 johria  staff    68B Aug  8 11:15 com.apple.tccd/
drwx------@  2 johria  staff    68B Aug  8 13:16 com.apple.tonelibraryd/
drwx------@  2 johria  staff    68B Aug  8 11:15 com.docker.helper/
-rw-------   1 johria  staff   377K Aug  8 12:01 com.dropbox.DropboxMacUpdate.33.3.14.dmg
-rw-------   1 johria  staff   100M Aug  8 12:01 com.getdropbox.dropbox.33.3.14.dmg
drwx------@  2 johria  staff    68B Aug  8 11:15 com.getdropbox.dropbox.garcon/
drwx------@  3 johria  staff   102B Aug  8 13:24 com.microsoft.Office365ServiceV2/
drwx------@  5 johria  staff   170B Aug  8 13:24 com.microsoft.Outlook/
drwxr-xr-x   7 johria  staff   238B Aug  8 11:15 hsperfdata_johria/
drwxr-xr-x   2 johria  staff    68B Aug  8 11:15 jetty-0.0.0.0-8983-webapp-_solr-any-7515977943968857408.dir/
drwxr-xr-x   2 johria  staff    68B Aug  8 11:16 jetty-127.0.0.1-7474-browser-_browser-any-6981420697006606178.dir/
drwxr-xr-x   2 johria  staff    68B Aug  8 11:15 jna--1154555289/
drwxr-xr-x   3 johria  staff   102B Aug  8 20:09 shared-pasteboard/
drwxr-xr-x   4 johria  staff   136B Aug  9 08:30 sp_update/
drwx------   2 johria  staff    68B Aug  8 12:27 ssh-VRtvsp23Sc5t/
-rw-------   1 johria  staff   165B Aug  8 11:15 start_1172299703324100109.properties
-rw-------   1 johria  staff   536B Aug  8 11:15 xcrun_db
```

**Comment by minrk** [comment]
Strange. That looks like you do have permission to set permissions on /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T, which is what's failing above. Can you run:

```python
import os
import stat
path = '/var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T'
permissions = os.stat(path).st_mode
new_permissions = permissions | stat.S_ISVTX
os.chmod(path, new_permissions)
```

**Comment by AlJohri** [comment]
```
acc-16249 # python
Python 3.6.2 (default, Jul 25 2017, 11:20:33)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> import stat
>>> path = '/var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T'
>>> permissions = os.stat(path).st_mode
>>> new_permissions = permissions | stat.S_ISVTX
>>> os.chmod(path, new_permissions)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
PermissionError: [Errno 1] Operation not permitted: '/var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T'
```

**Comment by minrk** [comment]
And what about:

    stat /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T

?

**Comment by AlJohri** [comment]
```
$ stat /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T
16777220 437605 drwxr-xr-x 6 johria staff 0 204 "Aug  9 08:35:15 2017" "May 28 10:24:36 2017" "May 28 10:24:36 2017" "May 28 09:27:33 2017" 4096 0 0x100000 /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn
16777220 437608 drwx------ 78 johria staff 0 2652 "Aug  9 08:35:33 2017" "Aug  9 09:49:33 2017" "Aug  9 09:49:33 2017" "May 28 09:27:33 2017" 4096 0 0x100000 /var/folders/c5/sxpknfp571v3ydglf4305g9m0000gn/T
````

**Comment by minrk** [comment]
Weird! All evidence points to you having permission to set the bit, I don't know why it wouldn't work. In any case, #286 should suppress the warning in this particular case.

**Pull request merged**